annotated.go

```diff
@@ -421,61 +428,66 @@ func (ann *annotated) results() (
 			continue
 		}
 
-		field := reflect.StructField{
-			Name: fmt.Sprintf("Field%d", i),
-			Type: t,
-		}
-
-		if i < len(ann.As) {
-			if !t.Implements(ann.As[i]) {
-				return nil, nil, fmt.Errorf("invalid fx.As: %v does not implement %v", t, ann.As[i])
+		for j := 0; j < numStructs; j++ {
+			field := reflect.StructField{
+				Name: fmt.Sprintf("Field%d", i),
+				Type: t,
 			}
-			field.Type = ann.As[i]
-		}
 
-		if i < len(ann.ResultTags) {
-			field.Tag = reflect.StructTag(ann.ResultTags[i])
+			if len(ann.As) > 0 && i < len(ann.As[j]) {
+				if !t.Implements(ann.As[j][i]) {
+					return nil, nil, fmt.Errorf("invalid fx.As: %v does not implement %v", t, ann.As[i])
+				}
+				field.Type = ann.As[j][i]
+			}
+			if i < len(ann.ResultTags) {
+				field.Tag = reflect.StructTag(ann.ResultTags[i])
+			}
+			offsets[j][i] = len(outFields[j])
+			outFields[j] = append(outFields[j], field)
 		}
+	}
 
-		offsets[i] = len(outFields)
-		outFields = append(outFields, field)
+	var resTypes []reflect.Type
+	for i := 0; i < numStructs; i++ {
+		resTypes = append(resTypes, reflect.StructOf(outFields[i]))
 	}
 
-	outType := reflect.StructOf(outFields)
-	types = []reflect.Type{outType}
 	if hasError {
-		types = append(types, _typeOfError)
+		resTypes = append(resTypes, _typeOfError)
 	}
 
-	return types, func(results []reflect.Value) []reflect.Value {
-		out := reflect.New(outType).Elem()
-
+	return resTypes, func(results []reflect.Value) []reflect.Value {
 		var outErr error
-		for i, r := range results {
-			if i == len(results)-1 && hasError {
-				// If hasError and this is the last item,
-				// we are guaranteed that this is an error
-				// object.
-				if err, _ := r.Interface().(error); err != nil {
-					outErr = err
+		var remappedResults []reflect.Value
+
+		for structNum := 0; structNum < numStructs; structNum++ {
+			out := reflect.New(resTypes[structNum]).Elem()
+			for i, r := range results {
+				if i == len(results)-1 && hasError {
+					// If hasError and this is the last item,
+					// we are guaranteed that this is an error
+					// object.
+					if err, _ := r.Interface().(error); err != nil {
+						outErr = err
+					}
+					continue
```

Comment:
I think this nested loop could be inverted.
Right now, we iterate through all results for each output struct,
and try to check the error for each output struct.

If we invert the nested loop so that `range` results is outside and `range resTypes` is inside,
we'll check error only once,
and for each result, we'll assign different output fields as needed.
We'll have to populate remappedResults with `reflect.New(..)` in advance, though.
So something like:

```go
var outputs []reflect.Value  // final return values
for _, structType := range resTypes {
  outputs = append(outputs, reflect.New(structType).Elem())
}

for i, r := range results {
  if /* error case */ {
    // ...
  }

  // Fill each struct that cares about this field
  // with the value.
  for j, structType := range resTypes {
    if idx := offsets[j][i]; idx > 0 {
      outputs[j][i].Set(r)
    }
  }
}
```